vmm/task/src/main.rs

```diff
@@ -258,15 +292,36 @@ async fn init_vm_rootfs() -> containerd_shim::Result<()> {
     // For more information see https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
     tokio::fs::write("/sys/fs/cgroup/memory/memory.use_hierarchy", "1")
         .await
-        .unwrap();
+        .map_err(io_error!(e, "failed to set cgroup hierarchy to 1"))
+}
+
+// Continue to do initialization that depend on shared path.
+// such as adding guest hook, preparing sandbox files and namespaces.
+async fn late_init_call() -> Result<()> {
+    // Setup DNS, bind mount to /etc/resolv.conf
+    if Path::new(SANDBOX_DNS_PATH).exists() {
+        mount_static_mounts(DNS_MOUNTS.clone()).await?;
+    } else {
+        warn!(
+            "unable to find {} files in kuasar state dir",
+            SANDBOX_DNS_PATH
+        );
+    }
+
     Ok(())
 }
 
-async fn mount_static_mounts(mounts: Vec<StaticMount>) -> containerd_shim::Result<()> {
+async fn mount_static_mounts(mounts: Vec<StaticMount>) -> Result<()> {
     for m in mounts {
-        tokio::fs::create_dir_all(Path::new(m.dest))
-            .await
-            .map_err(io_error!(e, "failed to create {}: ", m.dest))?;
+        let dest = Path::new(m.dest);
+        // For DNS mounts, dest file already exist.
```

Comment:
this is a function of common purpose, has nothing to do with DNS mounts, and I think it is wrong to make decision by if the dest exists. maybe we should assume that `mount_static_mounts` only handle dest of type directory. those with file bind mount, you just call the `mount` directly.

Reply:
Updated